File: code/ex202.java
class Solution {
    public boolean isHappy(int n) {
        Set<Integer> visited = new HashSet<>();
        visited.add(n);
        while (n != 1) {
            n = getSum(n);
            if (visited.contains(n))    return false;
            visited.add(n);
        }
        return n == 1;
    }
    
    private int getSum(int num) {
        int sum = 0;
        while (num > 0) {
            sum += (num % 10) * (num % 10);
            num /= 10;
        }
        return sum;
    }
}

File: code/ex202.cpp
// Fast and slow pointers

class Solution {
public:
    bool isHappy(int n) {
        if (n == 1)    return 1;
        int fast = n, slow = n;
        while (1) {
            slow = getNext(slow);
            fast = getNext(getNext(fast));
            if (slow == 1)    return 1;
            if (slow == fast)    return 0;
        }
        return 0;
    }
    
private:
    int getNext(int n) {
        int sum = 0;
        while (n > 0) {
            sum += (n % 10) * (n % 10);
            n /= 10;
        }
        return sum;
    }
};